apps/x509.c

```diff
@@ -810,7 +810,7 @@ int x509_main(int argc, char **argv)
     if (CAfile == NULL) {
         if (!X509V3_set_issuer_pkey(&ext_ctx, privkey))
             goto end;
-        if (!cert_matches_key(x, privkey))
+        if (privkey != NULL && !cert_matches_key(x, privkey))
```

Comment:
Could we move this check inside the cert_matches_key?

Reply:
This check is already (indirectly) in there, making the function return 0 if privkey equals NULL.
Yet here we need to suppress the warning in this case.

Reply:
`cert_matches_key` looks like the apps-purposed wrapper around `X509_check_private_key` with the only purpose - raise a warning on mismatch. As we shouldn't raise this warning in absence of a private key, and, if we should get an error, we could call `X509_check_private_key` directly, i'd suggest returning 0 from `cert_matches_key` when privkey is not available.

Reply:
> `cert_matches_key` looks like the apps-purposed wrapper around `X509_check_private_key` with the only purpose - raise a warning on mismatch.

Not really - `cert_matches_key()` is a wrapper just for preventing spurious error queue entries.
Note that it is going to be used also in `apps/pkcs12.c` in a situation where a warning would not be appropriate.

> As we shouldn't raise this warning in absence of a private key, and, if we should get an error, we could call `X509_check_private_key` directly, i'd suggest returning 0 from `cert_matches_key` when privkey is not available.

As mentioned, `X509_check_private_key()` and thus `cert_matches_key()` already returns 0 in case the key parameter is NULL.
And this does not relieve us from adding the guard `privkey != NULL`, as is done now, 
since otherwise we would get an unsuitable warning which also confused your tests.

Reply:
BTW, @beldmit, did you add to, e.g., https://github.com/gost-engine/engine/blob/879e6111068536f27686223602cd1e4789331632/tcl_tests/ca.try#L61-L67
a hint that the expected output for these test cases is derived using, e..g, 
`openssl x509 -in $revoking_cert -noout -serial"`?

Reply:
You are right, will do.